File: clojure/anagram/src/anagram.clj
(ns anagram)

(defn- anagram? [word1 word2]
  (let [lower1 (clojure.string/lower-case word1)
        lower2 (clojure.string/lower-case word2)
        same-word (= lower1 lower2)
        same-anagram (= (sort lower1) (sort lower2))]
    (and same-anagram (not same-word))))

(defn anagrams-for [word candidates]
  (filter #(anagram? word %) candidates))

File: scala/anagram/src/main/scala/Anagram.scala


class Anagram(word: String) {
  def matches(candidates: Seq[String]) =
    candidates.filter { c => isAnagram(word, c) }

  def isAnagram(word: String, other: String) = {
    val w = word.toLowerCase
    val o = other.toLowerCase
    w != o && w.sorted == o.sorted
  }
}